File: scripts/generate-jlpt-examples.mjs
import { createClient } from "@supabase/supabase-js";

const NVIDIA_URL = "https://integrate.api.nvidia.com/v1/chat/completions";
const DEFAULT_MODEL = "meta/llama-3.1-8b-instruct";

function option(name, fallback) {
  const index = process.argv.indexOf(`--${name}`);
  return index >= 0 ? process.argv[index + 1] : fallback;
}

const level = Number(option("level", "5"));
const limit = Number(option("limit", "20"));
const offset = Number(option("offset", "0"));
const examplesPerWord = Number(option("examples-per-word", "2"));
const delay = Number(option("delay", "800"));
const dryRun = process.argv.includes("--dry-run");
const model = option("model", process.env.NVIDIA_MODEL || DEFAULT_MODEL);

if (!Number.isInteger(level) || level < 1 || level > 5) throw new Error("--level은 1부터 5까지 입력해 주세요.");
if (!Number.isInteger(limit) || limit < 1 || !Number.isInteger(offset) || offset < 0) throw new Error("--limit과 --offset을 확인해 주세요.");
if (!Number.isInteger(examplesPerWord) || examplesPerWord < 1) throw new Error("--examples-per-word는 1 이상이어야 합니다.");

const required = ["NVIDIA_API_KEY", "SUPABASE_URL", "SUPABASE_SERVICE_ROLE_KEY"];
for (const name of required) {
  if (!process.env[name]) throw new Error(`${name}가 .env.local에 없습니다.`);
}

const supabase = createClient(process.env.SUPABASE_URL, process.env.SUPABASE_SERVICE_ROLE_KEY, {
  auth: { persistSession: false },
});

function extractJson(text) {
  const cleaned = text.trim().replace(/^```(?:json)?\s*/i, "").replace(/\s*```$/, "");
  const start = cleaned.indexOf("{");
  const end = cleaned.lastIndexOf("}");
  if (start < 0 || end < start) throw new Error("모델 응답에서 JSON을 찾지 못했습니다.");
  return JSON.parse(cleaned.slice(start, end + 1));
}

async function generate(word, count) {
  const prompt = `한국인 JLPT 학습자를 위한 일본어 예문을 ${count}개 만드세요.

대상 단어: ${word.word}
읽기: ${word.furigana}
한국어 뜻: ${word.meaning}
JLPT 레벨: N${word.level}

조건:
- 모든 일본어 문장에 대상 단어 또는 자연스러운 활용형을 포함하세요.
- N${word.level} 수준에 적절한 현대 일본어를 사용하세요.
- 문장은 간결하고 서로 다른 상황이어야 합니다.
- reading은 문장 전체를 히라가나로 적고 문장부호를 유지하세요.
- 자연스럽고 정확한 한국어 번역을 제공하세요.
- 반드시 다음 형태의 JSON만 출력하세요.
{"examples":[{"japanese":"...","reading":"...","translation_ko":"..."}]}`;

  const response = await fetch(NVIDIA_URL, {
    method: "POST",
    headers: { Authorization: `Bearer ${process.env.NVIDIA_API_KEY}`, "Content-Type": "application/json" },
    body: JSON.stringify({
      model,
      messages: [
        { role: "system", content: "당신은 꼼꼼한 일본어 교사입니다. 유효한 JSON만 출력하세요." },
        { role: "user", content: prompt },
      ],
      temperature: 0.25,
      top_p: 0.9,
      max_tokens: 1200,
      stream: false,
    }),
  });
  if (!response.ok) throw new Error(`NVIDIA API ${response.status}: ${(await response.text()).slice(0, 500)}`);
  const payload = await response.json();
  const parsed = extractJson(payload.choices?.[0]?.message?.content || "");
  if (!Array.isArray(parsed.examples) || parsed.examples.length !== count) throw new Error(`예문이 정확히 ${count}개 생성되지 않았습니다.`);
  return parsed.examples.map((example) => {
    const row = {
      japanese: String(example.japanese || "").trim(),
      reading: String(example.reading || "").trim(),
      translation_ko: String(example.translation_ko || "").trim(),
    };
    if (!row.japanese || !row.reading || !row.translation_ko) throw new Error("예문 필수 항목이 비어 있습니다.");
    return row;
  });
}

const { data: words, error: wordsError } = await supabase
  .from("jlpt_vocabulary")
  .select("id, word, meaning, furigana, level")
  .eq("level", level)
  .order("id", { ascending: true })
  .range(offset, offset + limit - 1);
if (wordsError) throw wordsError;

console.log(`N${level} 단어 ${words.length}개 · ${model}`);
let succeeded = 0;
let skipped = 0;
let failed = 0;

for (const [index, word] of words.entries()) {
  const label = `[${index + 1}/${words.length}] ${word.word}`;
  try {
    let existing = 0;
    if (!dryRun) {
      const { count, error } = await supabase
        .from("jlpt_example_sentences")
        .select("id", { count: "exact", head: true })
        .eq("vocabulary_id", word.id);
      if (error) throw error;
      existing = count || 0;
      if (existing >= examplesPerWord) {
        skipped += 1;
        console.log(`${label}: 건너뜀 (${existing}개 저장됨)`);
        continue;
      }
    }

    const examples = await generate(word, examplesPerWord - existing);
    if (dryRun) console.log(JSON.stringify({ word: word.word, examples }, null, 2));
    else {
      const rows = examples.map((example) => ({ ...example, vocabulary_id: word.id, source: "nvidia_nim", model, is_verified: false }));
      const { error } = await supabase.from("jlpt_example_sentences").upsert(rows, { onConflict: "vocabulary_id,japanese", ignoreDuplicates: true });
      if (error) throw error;
      console.log(`${label}: ${examples.length}개 저장 (${existing + examples.length}개)`);
    }
    succeeded += 1;
  } catch (error) {
    failed += 1;
    console.error(`${label}: 오류 - ${error.message}`);
  }
  if (index < words.length - 1 && delay > 0) await new Promise((resolve) => setTimeout(resolve, delay));
}

console.log(`완료: 성공 ${succeeded}, 건너뜀 ${skipped}, 실패 ${failed}`);
if (failed > 0) process.exitCode = 1;

File: scripts/generate-jlpt-examples.py
#!/usr/bin/env python3
"""Generate JLPT example sentences with NVIDIA NIM and store them in Supabase."""

from __future__ import annotations

import argparse
import json
import os
import re
import ssl
import sys
import time
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path
from typing import Any


PROJECT_ROOT = Path(__file__).resolve().parents[1]
ENV_FILE = PROJECT_ROOT / ".env.local"
NVIDIA_BASE_URL = "https://integrate.api.nvidia.com/v1"
DEFAULT_MODEL = "meta/llama-3.1-8b-instruct"

# Python 3.13+ enables RFC 5280 strict validation by default. Some corporate
# HTTPS inspection certificates omit legacy X.509 extensions such as AKI.
# Relax only that compatibility check; CA, signature, and hostname validation
# remain enabled.
TLS_CONTEXT = ssl.create_default_context()
if hasattr(ssl, "VERIFY_X509_STRICT"):
    TLS_CONTEXT.verify_flags &= ~ssl.VERIFY_X509_STRICT

if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")
    sys.stderr.reconfigure(encoding="utf-8")


def load_env_file(path: Path) -> None:
    if not path.exists():
        return
    for raw_line in path.read_text(encoding="utf-8-sig").splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, value = line.split("=", 1)
        key = key.strip()
        value = value.strip().strip('"').strip("'")
        os.environ.setdefault(key, value)


def require_env(name: str) -> str:
    value = os.environ.get(name, "").strip()
    if not value:
        raise RuntimeError(f"{name} is missing. Add it to {ENV_FILE}.")
    return value


def request_json(
    url: str,
    *,
    method: str = "GET",
    headers: dict[str, str] | None = None,
    payload: Any | None = None,
    timeout: int = 300,
) -> Any:
    body = None if payload is None else json.dumps(payload, ensure_ascii=False).encode("utf-8")
    request = urllib.request.Request(url, data=body, headers=headers or {}, method=method)
    max_attempts = 5
    for attempt in range(1, max_attempts + 1):
        try:
            with urllib.request.urlopen(request, timeout=timeout, context=TLS_CONTEXT) as response:
                data = response.read().decode("utf-8")
                return json.loads(data) if data else None
        except urllib.error.HTTPError as error:
            details = error.read().decode("utf-8", errors="replace")
            is_retryable = error.code in {429, 500, 502, 503, 504, 529}
            if not is_retryable or attempt == max_attempts:
                raise RuntimeError(f"HTTP {error.code} from {url}: {details[:1000]}") from error
            retry_after = error.headers.get("Retry-After", "").strip()
            wait_seconds = int(retry_after) if retry_after.isdigit() else 5 * (2 ** (attempt - 1))
            print(
                f"Temporary API error ({error.code}). Retrying in {wait_seconds}s "
                f"({attempt}/{max_attempts})...",
                file=sys.stderr,
            )
            time.sleep(wait_seconds)
        except (TimeoutError, urllib.error.URLError) as error:
            reason = getattr(error, "reason", error)
            is_timeout = isinstance(reason, TimeoutError) or "timed out" in str(reason).lower()
            if not is_timeout or attempt == max_attempts:
                raise
            wait_seconds = 5 * (2 ** (attempt - 1))
            print(
                f"API read timed out. Retrying in {wait_seconds}s "
                f"({attempt}/{max_attempts})...",
                file=sys.stderr,
            )
            time.sleep(wait_seconds)

    raise RuntimeError(f"Request failed after {max_attempts} attempts: {url}")


def supabase_headers(service_key: str, *, prefer: str | None = None) -> dict[str, str]:
    headers = {
        "apikey": service_key,
        "Authorization": f"Bearer {service_key}",
        "Content-Type": "application/json",
    }
    if prefer:
        headers["Prefer"] = prefer
    return headers


def fetch_words(
    supabase_url: str,
    service_key: str,
    level: int,
    limit: int,
    offset: int,
) -> list[dict[str, Any]]:
    query = urllib.parse.urlencode(
        {
            "select": "id,word,meaning,furigana,level",
            "level": f"eq.{level}",
            "order": "id.asc",
            "limit": str(limit),
            "offset": str(offset),
        }
    )
    return request_json(
        f"{supabase_url.rstrip('/')}/rest/v1/jlpt_vocabulary?{query}",
        headers=supabase_headers(service_key),
    )


def fetch_existing_count(supabase_url: str, service_key: str, vocabulary_id: int) -> int:
    query = urllib.parse.urlencode(
        {"select": "id", "vocabulary_id": f"eq.{vocabulary_id}", "limit": "10"}
    )
    rows = request_json(
        f"{supabase_url.rstrip('/')}/rest/v1/jlpt_example_sentences?{query}",
        headers=supabase_headers(service_key),
    )
    return len(rows)


def extract_json(text: str) -> dict[str, Any]:
    cleaned = text.strip()
    cleaned = re.sub(r"^```(?:json)?\s*", "", cleaned, flags=re.IGNORECASE)
    cleaned = re.sub(r"\s*```$", "", cleaned)
    start = cleaned.find("{")
    end = cleaned.rfind("}")
    if start < 0 or end < start:
        raise ValueError("The model response did not contain a JSON object.")
    result = json.loads(cleaned[start : end + 1])
    if not isinstance(result, dict):
        raise ValueError("The model response JSON must be an object.")
    return result


def generate_examples(
    nvidia_key: str,
    model: str,
    word: dict[str, Any],
    count: int,
) -> list[dict[str, str]]:
    prompt = f"""Create {count} Japanese example sentences for a Korean JLPT learner.

Target word: {word['word']}
Reading: {word['furigana']}
Korean meaning: {word['meaning']}
JLPT level: N{word['level']}

Requirements:
- Every Japanese sentence must contain the target word or a natural inflected form of it.
- Use natural modern Japanese and vocabulary/grammar appropriate for N{word['level']}.
- Keep each sentence concise and make the situations meaningfully different.
- Reading must be the full sentence written in hiragana. Preserve Japanese punctuation.
- Korean translation must be accurate and natural.
- Return JSON only, with exactly this shape:
{{"examples":[{{"japanese":"...","reading":"...","translation_ko":"..."}}]}}
"""
    response = request_json(
        f"{NVIDIA_BASE_URL}/chat/completions",
        method="POST",
        headers={
            "Authorization": f"Bearer {nvidia_key}",
            "Content-Type": "application/json",
        },
        payload={
            "model": model,
            "messages": [
                {
                    "role": "system",
                    "content": "You are a meticulous Japanese language teacher. Output valid JSON only.",
                },
                {"role": "user", "content": prompt},
            ],
            "temperature": 0.25,
            "top_p": 0.9,
            "max_tokens": 1200,
            "stream": False,
            **(
                {"reasoning_effort": "none"}
                if model == "deepseek-ai/deepseek-v4-flash"
                else {}
            ),
            **(
                {"chat_template_kwargs": {"enable_thinking": False}}
                if model.startswith("qwen/qwen3.5-")
                else {}
            ),
        },
    )
    content = response["choices"][0]["message"]["content"]
    parsed = extract_json(content)
    examples = parsed.get("examples")
    if not isinstance(examples, list) or len(examples) != count:
        raise ValueError(f"Expected exactly {count} examples.")

    validated: list[dict[str, str]] = []
    for example in examples:
        if not isinstance(example, dict):
            raise ValueError("Each example must be an object.")
        values = {
            field: str(example.get(field, "")).strip()
            for field in ("japanese", "reading", "translation_ko")
        }
        if not all(values.values()):
            raise ValueError("An example is missing japanese, reading, or translation_ko.")
        validated.append(values)
    return validated


def save_examples(
    supabase_url: str,
    service_key: str,
    vocabulary_id: int,
    model: str,
    examples: list[dict[str, str]],
) -> None:
    rows = [
        {
            "vocabulary_id": vocabulary_id,
            "japanese": example["japanese"],
            "reading": example["reading"],
            "translation_ko": example["translation_ko"],
            "source": "nvidia_nim",
            "model": model,
            "is_verified": False,
        }
        for example in examples
    ]
    request_json(
        f"{supabase_url.rstrip('/')}/rest/v1/jlpt_example_sentences?on_conflict=vocabulary_id,japanese",
        method="POST",
        headers=supabase_headers(
            service_key,
            prefer="resolution=ignore-duplicates,return=minimal",
        ),
        payload=rows,
    )


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--list-models",
        action="store_true",
        help="List model IDs currently available to the configured NVIDIA API key.",
    )
    parser.add_argument("--level", type=int, choices=range(1, 6), default=5)
    parser.add_argument("--limit", type=int, default=20)
    parser.add_argument("--offset", type=int, default=0)
    parser.add_argument("--examples-per-word", type=int, default=2)
    parser.add_argument("--model", default=None)
    parser.add_argument("--delay", type=float, default=0.8)
    parser.add_argument("--dry-run", action="store_true")
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    if args.limit < 1 or args.examples_per_word < 1:
        raise ValueError("--limit and --examples-per-word must be positive.")

    load_env_file(ENV_FILE)
    nvidia_key = require_env("NVIDIA_API_KEY")
    if args.list_models:
        payload = request_json(
            f"{NVIDIA_BASE_URL}/models",
            headers={
                "Authorization": f"Bearer {nvidia_key}",
                "Accept": "application/json",
            },
        )
        models = payload.get("data", []) if isinstance(payload, dict) else []
        model_ids = sorted(
            str(item.get("id", "")).strip()
            for item in models
            if isinstance(item, dict) and item.get("id")
        )
        if not model_ids:
            raise RuntimeError("NVIDIA API returned no available model IDs.")
        print("Models currently available to this NVIDIA API key:")
        for model_id in model_ids:
            print(model_id)
        return 0

    supabase_url = require_env("SUPABASE_URL")
    service_key = require_env("SUPABASE_SERVICE_ROLE_KEY")
    model = args.model or os.environ.get("NVIDIA_MODEL", DEFAULT_MODEL)

    words = fetch_words(supabase_url, service_key, args.level, args.limit, args.offset)
    if not words:
        print("No vocabulary words matched the requested range.")
        return 0

    print(f"Generating {args.examples_per_word} examples for {len(words)} N{args.level} words with {model}")
    succeeded = 0
    skipped = 0
    failed = 0

    for index, word in enumerate(words, start=1):
        label = f"[{index}/{len(words)}] {word['word']}"
        try:
            existing = 0
            if not args.dry_run:
                existing = fetch_existing_count(supabase_url, service_key, word["id"])
                if existing >= args.examples_per_word:
                    skipped += 1
                    print(f"{label}: skipped ({existing} already stored)")
                    continue

            examples = generate_examples(
                nvidia_key,
                model,
                word,
                args.examples_per_word - existing,
            )
            if args.dry_run:
                print(json.dumps({"word": word["word"], "examples": examples}, ensure_ascii=False))
            else:
                save_examples(supabase_url, service_key, word["id"], model, examples)
                print(f"{label}: saved {len(examples)} ({existing + len(examples)} total)")
            succeeded += 1
        except Exception as error:  # Keep the batch moving and report each failed word.
            failed += 1
            print(f"{label}: ERROR: {error}", file=sys.stderr)
        if index < len(words) and args.delay > 0:
            time.sleep(args.delay)

    print(f"Done: {succeeded} succeeded, {skipped} skipped, {failed} failed")
    return 1 if failed else 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except KeyboardInterrupt:
        raise SystemExit("Interrupted.")
    except Exception as error:
        raise SystemExit(f"ERROR: {error}") from error
